asserts/database.go

```diff
@@ -24,23 +24,52 @@ package asserts
 import (
 	"errors"
 	"fmt"
-	"io/ioutil"
-	"net/url"
 	"os"
 	"path/filepath"
 	"regexp"
 	"strings"
 	"time"
-
-	"github.com/ubuntu-core/snappy/helpers"
 )
 
+// BuilderFromComps can build an assertion from its components.
+type BuilderFromComps func(headers map[string]string, body, content, signature []byte) (Assertion, error)
+
+// Backstore is a backstore for assertions. It can store and retrieve
+// assertions by type under primary paths (tuples of strings). Plus it
+// supports more general searches.
+type Backstore interface {
+	// Init initializes the backstore. It is provided with a function
+	// to build assertions from their components.
+	Init(buildAssert BuilderFromComps) error
+	// Put stores an assertion under the given unique primaryPath.
+	// It is responsible for checking that assert is newer than a
+	// previously stored revision.
+	Put(assertType AssertionType, primaryPath []string, assert Assertion) error
+	// Get loads an assertion with the given unique primaryPath.
+	// If none is present it returns ErrNotFound.
+	Get(assertType AssertionType, primaryPath []string) (Assertion, error)
+	// SearchByHeaders searches for assertions matching the given headers.
+	// It invokes foundCb for each found assertion.
+	// pathHint is an incomplete primary path pattern (with ""
+	// representing omitted components) that covers a superset of
+	// the results, it can be used for the search if helpful.
+	SearchByHeaders(assertType AssertionType, headers map[string]string, pathHint []string, foundCb func(Assertion)) error
+	// SearchBySuffix searches for assertions matching the given
+	// partial primary path without the last component plus
+	// suffixOfLast being a suffix of that last component.
+	// It invokes foundCb for each found assertion.
+	SearchBySuffix(assertType AssertionType, primaryPathWithoutLast []string, suffixOflast string, foundCb func(Assertion)) error
+}
+
 // DatabaseConfig for an assertion database.
 type DatabaseConfig struct {
-	// database backstore path
+	// database filesystem backstores path
 	Path string
```

Comment:
@niemeyer yea, wasn't unsure myself about that one, I would prefer to do that in a follow up branch though, it will need some unrelated tests changes as well